src/libcore/mem/maybe_uninit.rs

```diff
@@ -324,7 +324,7 @@ impl<T> MaybeUninit<T> {
     /// Gets a pointer to the contained value. Reading from this pointer or turning it
     /// into a reference is undefined behavior unless the `MaybeUninit<T>` is initialized.
     /// Writing to memory that this pointer (non-transitively) points to is undefined behavior
-    /// (except inside an `UnsafeCell<T>`).
+    /// (except inside an `UnsafeCell<T>` when already initialized).
```

Comment:
I don't really understand the goal of this change. It is also not correct; the interior mutability exception is entirely orthogonal to initialization.

Could you elaborate a bit more?

Reply:
Hm, maybe this change isn't helping to make things more clear...
This is roughly what I was doing:
```rust
let value: MaybeUninit<UnsafeCell<bool>> = MaybeUninit::uninit();
let slot = (&*value.as_ptr()).get() as *mut T;
slot.write(true);
```
But because `UnsafeCell::get` takes a reference, this would be invalid, right? There would temporary be a reference to an uninitialized `UnsafeCell<bool>`.

So I would think the exception that you can write through `MaybeUninit::as_ptr` when it wraps an `UnsafeCell` would only be correct if it was already initialized.

Reply:
> So I would think the exception that you can write through MaybeUninit::as_ptr when it wraps an UnsafeCell would only be correct if it was already initialized.

No, the exception always applies. It's just you cannot make use of it through `UnsafeCell::get`. When writing unsafe code, you have to get *all* clauses of all constructs you use satisfied. Your example is fine as far as `MaybeUninit::as_ptr` is concerned -- it it what you are doing later that breaks it. Under some proposed memory model, if you use `transmute` to turn `*const UnsafeCell<bool>` into `*mut bool`, that is allowed and then you may use `ptr:.write` to initialize the `bool`.